### ISSUE TITLE
List should have an overload of replaceSubrange(_:with:) that accepts its indices

### ISSUE THREAD

**Issue by bdash** [realm/realm-cocoa#4760]
With an array you can replace its contents by doing:
```swift
a.replaceSubrange(a.indices, with: [1, 2, 3])
```

With `List<T>` you get an error because `List.Indices` is `DefaultRandomAccessIndices<List<T>>`, for which there's no overload. This should really work.

I took a brief stab at making it work but ran into a mysterious problem.  [`RangeReplaceableCollection`](https://developer.apple.com/reference/swift/rangereplaceablecollection) provides some overloads that take a `CountableRange<Int>`, so I tried changing `List.Indices` to be an alias for `CountableRange<Int>`, but for some reason the Swift compiler still won't use that overload. It appears to only see the overloads that take a `Range<Int>`.

**Comment by bdash** [comment]
Note that this isn't really about the specific overload mentioned in the tile, but about the wider question as to why some of `RangeReplaceableCollection`'s overloads are not available on `List`. It may also apply to our other collection types.

**Comment by austinzheng** [comment]
As part of #5276, I wrote a bunch of unit tests that, for pre-Swift 4.0 builds, just test the built-in functionality. One of these tests looks like:

```
func testReplaceOpenSubrange() {
        let subrange: Range<Int> = 2..<5
        let newElements = [10, 11, 12, 13, 14, 15, 16]
        list.replaceSubrange(subrange, with: makeSwiftIntObjects(from: newElements))
        array.replaceSubrange(subrange, with: newElements)
        compare(array: array, with: list)
}
```
This test compiles and passes successfully for all Swift versions from 3.0 to 3.2 (Xcode 8.0 and onward).

**Comment by austinzheng** [comment]
Fixed by #5276. I talked with @bdash and we decided that we can handle any other methods if users ask for them.
